azurelinuxagent/ga/update.py

```diff
@@ -392,17 +436,58 @@ def _process_goal_state(self, protocol, exthandlers_handler, remote_access_handl
 
         try:
             if incarnation != self.last_incarnation:  # TODO: This check should be based in the etag for the extensions goal state
-                exthandlers_handler.run()  # TODO: The info for the extensions within run() should be based on the extensions goal state
+                if not self._extensions_summary.converged:
+                    message = "A new goal state was received, but not all the extensions in the previous goal state have completed: {0}".format(self._extensions_summary)
+                    logger.warn(message)
+                    add_event(op=WALAEventOperation.GoalState, message=message, is_success=False, log_event=False)
+                    if self._is_initial_goal_state:
+                        self._on_initial_goal_state_completed(self._extensions_summary)
+                self._extensions_summary = ExtensionsSummary()
+                exthandlers_handler.run()
 
             # report status always, even if the goal state did not change
             # do it before processing the remote access, since that operation can take a long time
-            exthandlers_handler.report_ext_handlers_status(incarnation_changed=incarnation != self.last_incarnation)
+            self._report_status(exthandlers_handler, incarnation_changed=incarnation != self.last_incarnation)
 
             if incarnation != self.last_incarnation:
                 remote_access_handler.run()
         finally:
             self.last_incarnation = incarnation
 
+    def _report_status(self, exthandlers_handler, incarnation_changed):
+        # report_ext_handlers_status does its own error handling and returns None if an error occurred
+        vm_status = exthandlers_handler.report_ext_handlers_status(incarnation_changed=incarnation_changed)
+        if vm_status is None:
+            return
```

Comment:
Should we modify the self._extension_summary in this case? or maybe log it once per incarnation or something?

Reply:
we want to log it every time something changes, and changes can happen on every call to exthandlers_handler.report_ext_handlers_status

note that this logging is more for debugging the goal state, it is not related to the initial goal state logic (though it build on the same infrastructure)